### PR TITLE
Loosen \(QRX,StridedMatOrVec) to \(QRX,AbstractMatOrVec)

### LINKED ISSUE #16520: Matrix division \ gets confused by UnitRange on RHS
If solving `A \ B` where `B` is a `UnitRange` type will return a nonsense vector if the dimensions don't match.

For example:
```jl
b = 1:100
ones(3, 2) \ b
```
doesn't return a dimension mismatch error the way
```jl
ones(3, 2) \ collect(b)
```
does.

Instead it gives back a nonsense vector of the length of the `UnitRange`.

### PR DESCRIPTION
**Pull request by andreasnoack**
...and fix a ton of ambiguity issues caused by this. This also cleans up real-complex solves.
Fixes #16520.

### CODE REVIEW

File: base/linalg/qr.jl [hunk 1]
```diff
@@ -519,29 +519,68 @@ function A_ldiv_B!(A::QRPivoted, B::StridedMatrix)
...
 
-function \{TA,Tb}(A::Union{QR{TA},QRCompactWY{TA},QRPivoted{TA}}, b::StridedVector{Tb})
-    S = promote_type(TA,Tb)
-    m,n = size(A)
-    m == length(b) || throw(DimensionMismatch("left hand side has $m rows, but right hand side has length $(length(b))"))
-    AA = convert(Factorization{S}, A)
-    if n > m
-        x = A_ldiv_B!(AA, [b; zeros(S, n - m)])
+# convenience methods
+## return only the solution of a least squares problem while avoiding to promote
```
Comment (tkelman) [inline base/linalg/qr.jl]: avoiding promoting

File: test/linalg/bunchkaufman.jl [hunk 2]
```diff
@@ -11,7 +11,7 @@ n = 10
 n1 = div(n, 2)
 n2 = 2*n1
 
-srand(1234321)
+srand(1234322)
```
Comment (tkelman) [inline test/linalg/bunchkaufman.jl]: what's the reason for this change?

Comment (andreasnoack) [inline test/linalg/bunchkaufman.jl]: I kept getting this error https://travis-ci.org/JuliaLang/julia/jobs/132724032#L1787-L1808 which I cannot reproduce locally and which should be completely unrelated to the PR so I wanted to see if anything changed with a new seed.

File: test/linalg/bunchkaufman.jl [hunk 3]
```diff
@@ -107,7 +107,7 @@ end
 
 # test example due to @timholy in PR 15354
 let
-    A = rand(6,5); A = A'*A
+    A = rand(6,5); A = complex(A'*A) # to avoid calling the real-lhs-complex-rhs method
```
Comment (tkelman) [inline test/linalg/bunchkaufman.jl]: why didn't that method work?

Comment (andreasnoack) [inline test/linalg/bunchkaufman.jl]: It works but the test is with `==` here, probably to make sure that exactly the same method is called. For lhs `Matrix{Float64}` and rhs `Vector{Complex{Float64}}` the we are using the special method that reinterprets the rhs to avoid complex arithmetic.

Comment (tkelman) [inline test/linalg/bunchkaufman.jl]: so the reinterpret method would only differ in roundoff?

Comment (andreasnoack) [inline test/linalg/bunchkaufman.jl]: Yes.
